waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -913,16 +915,21 @@
              (pos-int? refresh-delay-mins))
          (symbol? action-fn)]}
   (let [refresh! (-> action-fn utils/resolve-symbol deref)
+        auth-update-time-atom (atom nil)
         auth-update-task (fn auth-update-task []
-                           (if-let [auth-str' (refresh! context)]
+                           (when-let [auth-str' (refresh! context)]
+                             (reset! auth-update-time-atom (t/now))
                              (reset! k8s-api-auth-str auth-str')))]
     (assert (fn? refresh!) "Refresh function must be a Clojure fn")
     (auth-update-task)
-    (when refresh-delay-mins
-      (du/start-timer-task
-        (t/minutes refresh-delay-mins)
-        auth-update-task
-        :delay-ms (* 60000 refresh-delay-mins)))))
+    {:cancel-fn (if refresh-delay-mins
+                  (du/start-timer-task
+                    (t/minutes refresh-delay-mins)
+                    auth-update-task
+                    :delay-ms (* 60000 refresh-delay-mins))
+                  (constantly nil))
+     :query-state-fn (fn query-auth-token-state []
+                       {:last-token-retrieve-time @auth-update-time-atom})}))
```

Comment:
Maybe this function should just return the atom's contents (and the atom can have a more generic name). That way the implementation could decide all of the state it wants to expose... But I'm not sure how we'd achieve this since `action-fn` currently just returns a string? Maybe we could have it return a string or a 2-count sequence?

Reply:
We can make a backwards incompatible change and have the `action-fn` return a map instead.